keps/sig-apps/0032-portable-service-definitions.md

```diff
@@ -0,0 +1,143 @@
+# KEP: Portable Service Definitions
+
+---
+kep-number: 31
+title: Portable Service Definitions
+authors:
+  - "@mattfarina"
+owning-sig: sig-apps
+participating-sigs:
+  - sig-service-catalog
+reviewers:
+  - "@carolynvs"
+  - "@kibbles-n-bytes"
+  - "@duglin"
+  - "@jboyd01"
+  - "@prydonius"
+  - "@kow3ns"
+approvers:
+  - "@mattfarina"
+  - "@prydonius"
+  - "@kow3ns"
+editor: TBD
+creation-date: 2018-11-13
+last-updated: 2018-11-13
+status: provisional
+see-also:
+replaces:
+superseded-by:
+
+---
+
+# Portable Service Definitions
+
+## Table of Contents
+
+A table of contents is helpful for quickly jumping to sections of a KEP and for highlighting any additional information provided beyond the standard KEP template.
+[Tools for generating][] a table of contents from markdown are available.
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [User Stories [optional]](#user-stories-optional)
+      * [Story 1](#story-1)
+      * [Story 2](#story-2)
+    * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+## Summary
+
+The goal of this feature is to enable an application to be deployed into multiple environments while relying on external services that are not part of the application. The service instances may or may not be managed by Kubernetes. For example, take a WordPress application that relies on MySQL. If the application is running in GKE or AKS it may leverage a Google or Azure MySQL as a Service. If it is running in Minikube or in a bare metal cluster it may use MySQL managed by the Kubernetes cluster. In all of these cases, a resource is declared asking for the service (e.g., MySQL) and the credentials to use the service are returned in a common manner. This enables portability from one environment to another.
```

Comment:
If I understood the idea correctly, a similar existing example of the proposed pattern is Kubernetes Ingress resource, where there are multiple implementations of Ingress controllers (NGINX, Contour etc) all supporting the core "Ingress" resource, and extending supported features via annotations (https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/) or custom resources (https://github.com/heptio/contour/blob/master/docs/ingressroute.md) correspondingly.

It would be nice to compare this proposal to Ingress as an existing example known to many people with its own pros/cons, and take some ideas from it.
https://kubernetes.io/docs/concepts/services-networking/ingress/#ingress-controllers

Reply:
Ingress should serve as a warning:  As an abstract API it is the lowest-common-factor of all the implementations.  It is a pretty unsatisfying experience and probably is not long for this world in its current form.

Reply:
We actually talked about the Ingress model with annotations vs other models the other day. While I can't say I know the right way I will say we are going to take lessons learned from Ingress into account.